File: static/python/parity3.py
# Parity3.h 848813e190b1b85a8e75107fe8513c3be38ad1a9
import parity1

PRECOMPUTED_PARITY = [parity1.parity(i) for i in range(1 << 16)]


# @include
def parity(x):
    MASK_SIZE = 16
    BIT_MASK = 0xFFFF
    return (PRECOMPUTED_PARITY[x >> (3 * MASK_SIZE)] ^
            PRECOMPUTED_PARITY[(x >> (2 * MASK_SIZE)) & BIT_MASK] ^
            PRECOMPUTED_PARITY[(x >> MASK_SIZE) & BIT_MASK] ^
            PRECOMPUTED_PARITY[x & BIT_MASK])
# @exclude

File: static/java/Parity3.java
public class Parity3 {
  private static short[] precomputedParity;

  static {
    precomputedParity = new short[1 << 16];
    for (int i = 0; i < (1 << 16); ++i) {
      precomputedParity[i] = Parity1.parity(i);
    }
  }

  // @include
  public static short parity(long x) {
    final int MASK_SIZE = 16;
    final int BIT_MASK = 0xFFFF;
    // clang-format off
    return (short) (
        precomputedParity[(int)((x >>> (3 * MASK_SIZE)) & BIT_MASK)]
        ^ precomputedParity[(int)((x >>> (2 * MASK_SIZE)) & BIT_MASK)]
        ^ precomputedParity[(int)((x >>> MASK_SIZE) & BIT_MASK)]
        ^ precomputedParity[(int)(x & BIT_MASK)]);
    // clang-format on
  }
  // @exclude
}
